src/coprocessor/service.rs

```diff
@@ -0,0 +1,1083 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::mem;
+use futures::{future, stream, Future, Stream};
+use futures::sync::mpsc;
+use protobuf::{CodedInputStream, Message};
+
+use kvproto::{coprocessor as coppb, errorpb, kvrpcpb};
+use tipb::executor::ExecType;
+use tipb::select::DAGRequest;
+use tipb::analyze::{AnalyzeReq, AnalyzeType};
+use tipb::checksum::{ChecksumRequest, ChecksumScanOn};
+
+use util;
+use util::futurepool;
+use util::time::{self, Instant};
+use storage;
+use server::Config;
+use server::readpool::{self, ReadPool};
+
+use super::*;
+use super::util as cop_util;
+use super::dag::DAGContext;
+use super::statistics::analyze::AnalyzeContext;
+use super::checksum::ChecksumContext;
+use super::local_metrics::BasicLocalMetrics;
+use super::dag::executor::ExecutorMetrics;
+
+// If handle time is larger than the lower bound, the query is considered as slow query.
+const SLOW_QUERY_LOWER_BOUND: f64 = 1.0; // 1 second.
+
+const OUTDATED_ERROR_MSG: &str = "request outdated.";
+const COPROCESSOR_BUSY_ERROR_MSG: &str = "coprocessor is busy";
+
+pub struct Service {
+    engine: Box<storage::Engine>,
+    read_pool: ReadPool<ReadPoolContext>,
+    recursion_limit: u32,
+    batch_row_limit: usize,
+    stream_batch_row_limit: usize,
+    stream_channel_size: usize,
+    max_handle_duration: Duration,
+}
+
+impl Clone for Service {
+    fn clone(&self) -> Self {
+        Service {
+            engine: self.engine.clone(),
+            read_pool: self.read_pool.clone(),
+            ..*self
+        }
+    }
+}
+
+impl util::AssertSend for Service {}
+
+impl Service {
```

Comment:
I don't think `Service` is a good name. Now we have `kv`, `debug` and `import` `Service`s. All of them will listen on a port and start a gRPC service.

Reply:
Agree. Suggestions? Or keep previous EndPoint? Notice that the `Service` now is different to the `Endpoint` in previous. `Endpoint` is singleton while `Service` have multiple instances (each in a grpc thread).